### PR TITLE
[BD-05][TNL-7915][BB-3611] Improve ZIP File content structure

### PR DESCRIPTION
**Pull request by 0x29a**
This PR contains an improvement of the submission files archive structure, which can be generated by clicking this button in the "Data Download" section of the instructor dashboard:
image

Previous structure example:
```
.
└── CourseId
    ├── BlockId1
    │   ├── StudentId1
    │   │   ├── attachments
    │   │   │   ├── SomeFile1
    │   │   │   └── SomeFile2
    │   │   ├── part_0.txt
    │   │   └── part_1.txt
    │   └── StudentId2
    │       ├── part_0.txt
    │       └── part_1.txt
    ├── BlockId2
    │   └── StudentId3
    │       ├── attachments
    │       │   └── SomeFile4
    │       └── part_0.txt
    └── downloads.csv
```

New structure example:
```
.
├── [1]Some Section, [1]Some Subsection, [1]Unit
│   ├── [1] - 00f636b9ac6d480c9fb95c23bf1d2129 - prompt_0.txt
│   ├── [1] - 00f636b9ac6d480c9fb95c23bf1d2129 - prompt_1.txt
│   ├── [1] - edx - prompt_0.txt
│   ├── [1] - edx - prompt_1.txt
│   ├── [2] - edx - prompt_0.txt
│   └── [2] - edx - Structure and Interpretation of Computer Programs.pdf
├── [1]Some Section, [2]Some Subsection, [1]Unit
│   └── [1] - edx - prompt_0.txt
├── [1]Some Section, [2]Some Subsection, [2]Unit
│   ├── [1] - edx - prompt_0.txt
│   └── [1] - edx - the_most_dangerous_kitten.jpg
└── downloads.csv
```

**JIRA tickets**:
- [BB-3611](https://tasks.opencraft.com/browse/BB-3611).
- [TNL-7915](https://openedx.atlassian.net/browse/TNL-7915).
- [BLENDED-799](https://openedx.atlassian.net/browse/BLENDED-799).

**Sandboxes**:
- https://pr27218.sandbox.opencraft.hosting/

**Discussions**:
- [First Slack thread](https://openedx.slack.com/archives/C014VAUJUJ0/p1612272008003000).
- [Second Slack thread](https://openedx.slack.com/archives/C014VAUJUJ0/p1612968435011100).
- [Third Slack thread](https://openedx.slack.com/archives/C014VAUJUJ0/p1617028572003900).

**Dependencies**:
- https://github.com/edx/edx-platform/pull/27218 depends on this PR.

**Testing instructions**:

###### Step I: checkout edx-platform
1. `cd $(devstack_root)`
1. `cd ../edx-platform`
1. `git remote add opencraft https://github.com/open-craft/edx-platform`
1. `git fetch opencraft 0x29a/bb3611/improve_zip_file_content_structure`
1. `git checkout 0x29a/bb3611/improve_zip_file_content_structure`

###### Step II: install edx-ora2
1. `cd $(devstack_root)`
1. `cd ../src/`
1. `git clone https://github.com/edx/edx-ora2/`
1. `cd edx-ora2`
1. `git checkout 0x29a/bb3611/improve_zip_file_content_structure`
1. `make install-local-ora`
1. `cd $(devstack_root)`
1. `make lms-restart`

###### Step III: prepare course
1. Add ORA block to the demo course, enable optional file attachments.
1. Change name of subsection where ORA was added to:
```
*************************************************************************************************************************************************************************************************************************************************************** | THIS PART WILL BE TRUNCATED
```
3. Submit answer as `staff@example.com` user, attach file with a following name:
```
***************************************************************************************************************************************************************************************************************************** | THIS PART WILL BE TRUNCATED.jpg
```

###### Step IV: create external ID
1. `cd $(devstack_root)`
1. `make lms-shell`
1. `python manage.py lms shell`
1. `from openedx.core.djangoapps.external_user_ids.models import ExternalId`
1. `from django.contrib.auth import get_user_model`
1. `ExternalId.add_new_user_id(get_user_model().objects.get(email='staff@example.com'), 'mb_coaching')`

###### Step V: check zip structure
1. Go to the "Download Data" section of the instructor dashboard.
1. Click "Generate Submission Files Archive".
1. Download generated file.
1. Ensure that the subsection name is truncated, as well as filename base. Ensure that submission files for a `staff` user contain external ID.

**Reviewers**
- [ ] @giovannicimolin 
- [ ] @jnlapierre

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @0x29a! I've created [BLENDED-799](https://openedx.atlassian.net/browse/BLENDED-799) to keep track of it in Jira.   More details are on the [BD-05](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1529970851/BD-05+ORA+v3+Usability+Moderation) project page.


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

**Review by giovannicimolin — COMMENTED** [review]
@0x29a I'm testing this with a test course which has multiple ORA and some units/sub-units to test possible edge cases.

Here's the course export: 
[course.19o_d71t.tar.gz](https://github.com/edx/edx-ora2/files/6250522/course.19o_d71t.tar.gz)

And this is a ZIP file export generated for this course:
**With anonymous ID:**
[OpenCraft_Test_tests_submission_files_2021-04-02-1544.zip](https://github.com/edx/edx-ora2/files/6250527/OpenCraft_Test_tests_submission_files_2021-04-02-1544.zip)

**With username::**
[OpenCraft_Test_tests_submission_files_2021-04-02-1542.zip](https://github.com/edx/edx-ora2/files/6250529/OpenCraft_Test_tests_submission_files_2021-04-02-1542.zip)

----

So far things seem okay. I'll keep digging the course and implementation.

**Review by giovannicimolin — CHANGES_REQUESTED** [review]
@0x29a I found some issues with the export.

Using this course (expanded from the first one I've posted above):  [course.830itub8.tar.gz](https://github.com/edx/edx-ora2/files/6250614/course.830itub8.tar.gz)

I found and issue with the filename that's causing these issues:
1. The submissions in the second folder (`[1]Section 1, [2]Subsection 2, [2]Unit`) start with index 2, even when there's a single ORA in the unit.
2. The submissions in the first folder (`[1]Section 1, [2]Subsection 2, [1]Unit` - this is a unit with 3 ORAs) have all the same index (`1`). This is causing a name clash issue and not all submissions are being exported properly. I'm submitting the assessments with 2 users.
  * Expected for `[1]Section 1, [2]Subsection 2, [1]Unit`: 6 files (2 for each ORA).
  * What's comes in the ZIP file: 
image
  * What's extracted (files with the same name cause filesystem conflicts):
image

Additionally, the `CSV` file contents aren't very clear. Is there a real need for it? Was it in the original requirements? If not, we can probably remove it.

**Comment by 0x29a** [comment]
@giovannicimolin, thanks for the review.

In https://github.com/edx/edx-ora2/pull/1610/commits/7c45194ceab9598a8d50894c23a76729ed42916c commit I fixed typo in `_submission_filepath` function that was leading to incorrect ORA indexes in filename: `ora_path_info["unit_index"]` -> `ora_path_info["ora_index"]`.

Also, I added block type check here to prevent counting non-ORA blocks:
https://github.com/edx/edx-ora2/blob/7c45194ceab9598a8d50894c23a76729ed42916c/openassessment/data.py#L974-L984

And finally, I added handling of two edge cases:
1. When ORA block was removed from course, but submission is still there.
2. When section/sub-section/unit is empty.

File: openassessment/data.py [hunk 1]
```diff
@@ -959,36 +952,204 @@ def _download_file_by_key(cls, key):
...
+                                "section_name": section_name,
+                                "sub_section_index": sub_section_index,
+                                "sub_section_name": sub_section_name,
+                                "unit_index": unit_index,
+                                "unit_name": unit_name,
+                                "ora_index": block_index,
+                                "ora_name": block_name,
+                            }
+
+        return path_info
```
Comment (0x29a) [inline openassessment/data.py:986]: Note: this wasn't covered with tests because it relies on platform internals.

File: openassessment/data.py [hunk 2]
```diff
@@ -959,36 +952,204 @@ def _download_file_by_key(cls, key):
...
+                        if _usernames_enabled()
+                        else "anonymoususerid__anonymous_user_id"
+                    ),
+                    output_field=CharField(),
+                ),
+            )
+            .values("student_id", "path_id")
+        )
+
+        return {user["student_id"]: user["path_id"] for user in users}
```
Comment (0x29a) [inline openassessment/data.py:1026]: Note: this wasn't covered with tests as well, since it relies on `ExternalId` model from Open edX core.

**Review by giovannicimolin — APPROVED** [review]
:+1:

- [x] I tested this:
1. Checked out this branch along with the https://github.com/edx/edx-platform/pull/27218.
2. Create a course structure that trigger edge cases with multiple ORAs. Export here: 
[course.7nx5cez9.tar.gz](https://github.com/edx/edx-ora2/files/6258759/course.7nx5cez9.tar.gz)
3. Submitted responses to the ORAs using `edx` and `verified` users. 
4. Exported the ZIP file (no PII - default): 
[OpenCraft_Test_tests_submission_files_2021-04-05-1352.zip](https://github.com/edx/edx-ora2/files/6258765/OpenCraft_Test_tests_submission_files_2021-04-05-1352.zip)
5. Enabled `ENABLE_ORA_USERNAMES_ON_DATA_EXPORT` and exported the same file again: 
[OpenCraft_Test_tests_submission_files_2021-04-05-1424.zip](https://github.com/edx/edx-ora2/files/6258773/OpenCraft_Test_tests_submission_files_2021-04-05-1424.zip)
- [x] I read through the code
- [ ] ~~I checked for accessibility issues~~ NA
- [x] Includes documentation: docstrings. User documentation coming up on follow-up ticket.

@0x29a LGTM. :+1: 
Can you check why the sandbox is failing? https://manage.opencraft.com/instance/25930/

@jnlapierre @mattcarter This is ready for review.

**Comment by 0x29a** [comment]
@giovannicimolin
> Can you check why the sandbox is failing? https://manage.opencraft.com/instance/25930/

Most likely, it's because I rebased the branch right after PR creation and Ocim was trying to check out to commit that wasn't a part of git tree. I updated it and created a new app server.

**Review by jnlapierre — APPROVED** [review]
LGTM and nice job cleaning up and refactoring some of the code to shared methods.

**Review by jnlapierre — APPROVED** [review]
LGTM
